quickwit/quickwit-serve/src/pprof.rs

```diff
@@ -0,0 +1,159 @@
+        // A new run will overwrite the flamegraph_data
+        flamegraph_data: Option<Vec<u8>>,
+    }
+
+    let profiler_state = Arc::new(Mutex::new(ProfilerState {
+        profiler_guard: None,
+        flamegraph_data: None,
+    }));
+
+    #[derive(Deserialize)]
+    struct ProfilerQueryParams {
+        max_duration: Option<u64>, // max allowed value is 300 seconds, default is 30 seconds
+        sampling: Option<i32>,     // max value is 1000, default is 100
+    }
+
+    let start_profiler = {
+        let profiler_state = Arc::clone(&profiler_state);
+        warp::path!("pprof" / "start")
+            .and(warp::query::<ProfilerQueryParams>())
+            .and_then(move |params: ProfilerQueryParams| {
+                start_profiler_handler(profiler_state.clone(), params)
+            })
+    };
+
+    let stop_profiler = {
+        let profiler_state = Arc::clone(&profiler_state);
+        warp::path!("pprof" / "flamegraph")
+            .and_then(move || stop_profiler_handler(Arc::clone(&profiler_state)))
+    };
+
+    async fn start_profiler_handler(
+        profiler_state: Arc<Mutex<ProfilerState>>,
+        params: ProfilerQueryParams,
+    ) -> Result<impl warp::Reply, warp::Rejection> {
+        let mut state = profiler_state.lock().unwrap();
+        if state.profiler_guard.is_none() {
+            let max_duration = params.max_duration.unwrap_or(30).min(300);
+            let sampling = params.sampling.unwrap_or(100).min(1000);
+            state.profiler_guard = Some(pprof::ProfilerGuard::new(sampling).unwrap());
+            let profiler_state = Arc::clone(&profiler_state);
+            tokio::spawn(async move {
+                time::sleep(Duration::from_secs(max_duration)).await;
+                save_flamegraph(profiler_state).await;
+            });
+            Ok(warp::reply::with_status(
+                "CPU profiling started",
+                warp::http::StatusCode::OK,
+            ))
+        } else {
+            Ok(warp::reply::with_status(
+                "CPU profiling is already running",
+                warp::http::StatusCode::BAD_REQUEST,
+            ))
+        }
+    }
+
+    async fn stop_profiler_handler(
+        profiler_state: Arc<Mutex<ProfilerState>>,
+    ) -> Result<impl warp::Reply, warp::Rejection> {
+        save_flamegraph(profiler_state.clone()).await;
+
+        let state = profiler_state.lock().unwrap();
+        if let Some(data) = state.flamegraph_data.clone() {
+            Ok(warp::reply::with_header(data, "Content-Type", "image/svg+xml").into_response())
+        } else {
+            Ok(warp::reply::with_status(
+                "No flamegraph available",
+                warp::http::StatusCode::BAD_REQUEST,
+            )
+            .into_response())
+        }
+    }
+
+    async fn save_flamegraph(
+        profiler_state: Arc<Mutex<ProfilerState>>,
+    ) -> tokio::task::JoinHandle<()> {
+        spawn_blocking(move || {
```

Comment:
Technically it is better to use `quickwit_common:run_cpu_intensive`
because spawn blocking is an infinite pool.

That being said, here, I think we don't care much. 
Fix if you want.